torch_em/loss/wrapper.py

```diff
@@ -43,30 +44,72 @@ def forward(self, prediction, target, **kwargs):
 
 
 class ApplyMask:
-    def __call__(self, prediction, target, mask):
-        mask.requires_grad = False
+    def _crop(prediction, target, mask, channel_dim):
+        mask = mask.type(torch.bool)
+        pred = []
+        trgt = []
+        n_channels = prediction.shape[channel_dim]
+        # for now only support same mask for all channels
+        # to be able to stack/concat afterwards
+        # TODO: make this more efficient, and for different
+        # masks per channel
+        assert mask.shape[channel_dim] == 1
+        for c in range(n_channels):
+            p = torch.index_select(prediction, channel_dim, torch.LongTensor([c]))
+            t = torch.index_select(target, channel_dim, torch.LongTensor([c]))
+            pred.append(p[mask])
+            trgt.append(t[mask])
+        # transpose to have N x C
+        # needed in torch_em.loss.dice.flatten_samples
+        pred = torch.stack(pred).transpose(0, 1)
+        trgt = torch.stack(trgt).transpose(0, 1)
+        return pred, trgt
+
+    def _multiply(prediction, target, mask, channel_dim):
         prediction = prediction * mask
         target = target * mask
         return prediction, target
 
+    MASKING_FUNCS = {
+        "crop": _crop,
+        "multiply": _multiply,
+    }
+
+    def __init__(self, masking_method="crop", channel_dim=1):
+        if masking_method not in self.MASKING_FUNCS.keys():
+            raise ValueError(f"{masking_method} is not available, please use one of {list(self.MASKING_FUNCS.keys())}.")
+        self.masking_func = self.MASKING_FUNCS[masking_method]
+        self.channel_dim = channel_dim
+
+        self.init_kwargs = {
+            "masking_method": masking_method,
+            "channel_dim": channel_dim,
+        }
+
+    def __call__(self, prediction, target, mask):
+        mask.requires_grad = False
+        return self.masking_func(prediction, target, mask, self.channel_dim)
+
 
-class ApplyAndRemoveMask:
+class ApplyAndRemoveMask(ApplyMask):
```

Comment:
+1 for the inheritance (here and in the other classes), but I think we should then also introduce a `__init__` here that calls the super constructor.

Reply:
do you mean for better readibility? Because __init__ of the base class is called anyway if we do not overwrite it.

Reply:
Yea, you're right; I didn't think about this one long enough ;).